Add MTA Plugin

Jasper-ketelaar opened the pull request:
# Description
This plugin adds overlays to three of the four rooms in Mage Training Arena. It supports Telekinetic Theatre, Alchemists' Playground and Creature Graveyard. The plugin has hint arrows for three of the four rooms to indicate where things are clearly and make it easier to see. These can be turned off for the whole plugin in the configuration.

## Telekinetic Theatre
The Telekinetic room overlay shows you on which side of the maze to stand and it shows you when you are able to cast the telekinetic grab spell correctly. The colours of the tiles that it indicates for you to stand on turn green when you are able to cast the spell.

![telekinetic3](https://user-images.githubusercontent.com/13006588/40925027-f945d7ce-6818-11e8-9a23-3d726ccec89e.gif)

The plugin also offers the ability to instead of one tile draw the line of tiles on which you are able to stand to correctly cast the spell.

![teliklines](https://i.imgur.com/gu04mrg.png)

## Alchemists' Playground
The Alchemy room overlay shows you in what cupboard what item is based on the items you have previously gotten. This room operates on rotations and the order of the items in the cupboard is always the same. This allows you to devise an item's location based on other items you have found. The overlay will also highlight items in your inventory that give the most points

![alchemy3](https://user-images.githubusercontent.com/13006588/40924677-302b7f60-6818-11e8-90bc-35d04438ba54.gif)

The plugin also allows you to change from icons to text

![alchtext](https://i.imgur.com/vGbKL90.png)

## Creature Graveyard
The Graveyard room is a lot less complex than the other two. This overlay will add a counter infobox displaying how many fruit your bones will get you and the colour changes depending on whether this is enough for a point. The games bones widget is hidden as it is rendered usesless by this plugin and may interfere with the counter boxes.

![graveyardcounter](https://i.imgur.com/fTH2Obb.png)

This room also has an overlay for your inventory that turns green when the points of the bones in your inventory are enough to get a graveyard point. This overlay is off by default and can be turned on from the config.

![graveyard2](https://user-images.githubusercontent.com/13006588/40925312-be22db78-6819-11e8-9a15-8d052a89ccb8.gif)

## Enchanting Chamber

Has a simple counter for the enchantment room showing how many dragonstones have spawned. This plugin will also hide the widgets of the room since they are not used by anyone who does this room. The room also uses hint arrows to indicate where the dragonstones are.

![enchantment](https://user-images.githubusercontent.com/13006588/40925466-1240cf6c-681a-11e8-9839-2d679d329db4.gif)


## Settings
The settings allow you to turn off each supported room individually and they allow you to turn off the filling of clickboxes for items in the Graveyard and Alchemy room or turn off the hint arrows for all of the rooms:

![settings](https://i.imgur.com/Tr6YcuC.png) ![fillon](https://i.imgur.com/1weoybA.png) ![filloff](https://i.imgur.com/1vBCtxS.png)

forsco commented:
Not sure if I'm a fan of leaving the overlays rendered if the inventory isn't the active widget.

MagicfTail commented:
How about changing the boxes to outlines similar to how the Jewelry enhance plugins does it. It looks way better than having large coloured boxes

Jasper-ketelaar commented:
That is already in the configs @MagicfTail as outlined by the post. @forsco I'll add a config option for that.

deathbeam commented:
Config option? It should not even be a feature. Also, what about https://github.com/runelite/runelite/pull/3436? I like the icons there a lot more than the text and click boxes you have in alchemy room.

Jasper-ketelaar commented:
I like being able to see the clickboxes without my inventory open though, makes you able to tell where you need to click to alch before opening the inventory essentially making it faster. I can add it as a default off config option unless you mean it shouldn't be a feature as it is against game integrity? 

For the icons I'll add that.

deathbeam commented:
Not really game integrity, but stuff like this should be solved in different way (e.g alternate inventory, that should be completely separate feature, and there was already PR for that).

psikoi commented:
Also, I don't thin you need to be at the center of one of the sides for the grab to work, you just need to be anywhere on that side of the maze, can you highlight a bunch of tiles instead? the whole row? that would require less running around

Jasper-ketelaar commented:
Yeah you're right. The highlighting will turn green as soon as you are able to cast the spell the center tile was just for an indication of the side. I can make it highlight the whole row instead.

Jasper-ketelaar commented:
# Updated alchemy room with suggestions
Now has an option to enable icons instead of text which by default is on. Made the unknown cupboards be represented by the cake of guidance (question mark cake) icon. The alchemy room now also has a reset timer that indicates how long until the cupboards reset.

![alchemy2](https://user-images.githubusercontent.com/13006588/40914855-2533a538-67fa-11e8-8927-df8ff3b2f4cb.gif)

Jasper-ketelaar commented:
# Updated telekinetic room
With the suggestions of @psikoi I have now made the telekinetic room have a line of tiles to stand on instead of just one tile and the hint arrow now goes to the nearest tile to stand on. Updated main post with new gifs as well.

![telekinetic2](https://user-images.githubusercontent.com/13006588/40916573-874c8144-6800-11e8-99a5-55c19b5ec92c.gif)

Jasper-ketelaar commented:
# Enchantment room counter
Added a simple counter for the enchantment room showing how many dragonstones have spawned. This will also hide the widgets of the room since they are not used by anyone who does this room. 
![enchantment](https://i.imgur.com/Kyipm9M.png)

Jasper-ketelaar commented:
## Added graveyard room bones counter
Added an infobox counter for graveyard room which's text colour will change depending on how many points your bones will give you. The other overlay that goes over the inventory is now by default off but can still be toggled on.

![graveyardcounter](https://i.imgur.com/fTH2Obb.png)

psikoi commented:
I'm really liking this so far, if I could nit-pick for a bit:´

Instead of highlightinting the guarding model, why not just the tile beneath him? I think it would look cleaner.

Why are you highlighting the cupboard models? you already have item indicators on top of them, I don't think people need more.

Only complain I have about graveyard would probably be the inventory slots being rendered when you're in the spellbook.

**Mind you, I may have misunderstood some of the functionalities for this plugin, so some of the things I said may be wrong.**

Jasper-ketelaar commented:
@psikoi I've removed highlighting cupboard models and replaced it with a hint arrow. Added a global config to disable hint arrows as well for those who are not fans of them. The graveyard issue has been fixed in a previous commit I should update the images on the main post to reflect this. As for highlighting the tile instead of the guardian model, this has also been changed.

Jasper-ketelaar commented:
Updated the main post with new images, explanation and gifs to reflect the latest version of the plugin.

psikoi commented:
Looks really really good after those updates, one last thing I guess, for the alchemy room, when showing text instead of icons, can you find a way to use the text with a shadow? easier to read

Jasper-ketelaar commented:
Agreed, it's been changed to this
![newtextalch](https://i.imgur.com/sTlTCOu.png)

psikoi commented on runelite-api/src/main/java/net/runelite/api/coords/Direction.java:
```diff
@@ -47,5 +47,62 @@
+			case 0:
+			case 7:
+				return SOUTH;
+
```
remove empty line

psikoi commented on runelite-api/src/main/java/net/runelite/api/coords/Direction.java:
```diff
@@ -47,5 +47,62 @@
+			case 1:
+			case 2:
+				return WEST;
+
```
remote empty line

psikoi commented on runelite-api/src/main/java/net/runelite/api/coords/Direction.java:
```diff
@@ -47,5 +47,62 @@
+			case 3:
+			case 4:
+				return NORTH;
+
```
remote empty line

psikoi commented on runelite-api/src/main/java/net/runelite/api/coords/Direction.java:
```diff
@@ -47,5 +47,62 @@
+			case 5:
+			case 6:
+				return EAST;
+
```
remote empty line

psikoi commented on runelite-api/src/main/java/net/runelite/api/coords/Direction.java:
```diff
@@ -47,5 +47,62 @@
+		{
+			case NORTH:
+				return SOUTH;
+
```
remove empty line

psikoi commented on runelite-api/src/main/java/net/runelite/api/coords/Direction.java:
```diff
@@ -47,5 +47,62 @@
+
+			case SOUTH:
+				return NORTH;
+
```
remove empty line

psikoi commented on runelite-api/src/main/java/net/runelite/api/coords/Direction.java:
```diff
@@ -47,5 +47,62 @@
+
+			case EAST:
+				return WEST;
+
```
remove empty line

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyItem.java:
```diff
@@ -0,0 +1,102 @@
+
+public enum AlchemyItem
+{
+
```
remove empty line

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyItem.java:
```diff
@@ -0,0 +1,102 @@
+		this.name = name;
+	}
+
+	public static AlchemyItem forName(String name)
```
wouldn't fromName make more sense?

Jasper-ketelaar replied:
Yeah agreed. The code wasn't being used however so I decided to just remove the method.

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyItem.java:
```diff
@@ -0,0 +1,102 @@
+	public static int indexOf(String item)
+	{
+		AlchemyItem[] items = values();
+		for (int i = 0; i < items.length; i++)
```
This is me being over the top, but maybe add a line between the array and the for loop?

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyItem.java:
```diff
@@ -0,0 +1,102 @@
+		return -1;
+	}
+
+	public int getId()
```
Use the @Getter annotation on the id variable instead of creating actual getters

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyItem.java:
```diff
@@ -0,0 +1,102 @@
+		return id;
+	}
+
+	public String getName()
```
Use the @Getter annotation on the name variable instead of creating actual getters

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyItem.java:
```diff
@@ -0,0 +1,102 @@
+	public String toString()
+	{
+		String text = getName();
+		if (text.equals("helm"))
```
Again, being over the top, but can you add a line here? between the variable and the if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+	private final AlchemyItem[] locations = new AlchemyItem[CAPACITY];
+	private final List<GameObject> cupboards = new ArrayList<>(CAPACITY);
+	private final MTAPlugin plugin;
+	@Inject
```
I would add empty lines separating some of these variables, not everyone in this project does it, and is not required for the code style, but makes the code easier to read

Jasper-ketelaar replied:
Grouped variables by initialization time

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+		this.plugin = plugin;
+	}
+
+
```
Remove empty line

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+				infoBoxManager.removeIf(e -> e instanceof AlchemyRoomTimer);
+				infoBoxManager.addInfoBox(new AlchemyRoomTimer(plugin));
+			}
+			best = getBest();
```
Add line above

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+		if (!inside())
+			return;
+
+		GameObject spawn = event.getGameObject();
```
Add line between variable and the for loop

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+		for (int i = 0; i < CUPBOARDS.size(); i++)
+		{
+			int id = CUPBOARDS.get(i);
+			if (spawn.getId() == id || spawn.getId() - 1 == id)
```
Add line between variable and the if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+			return;
+
+		String message = wrapper.getMessage();
+		if (wrapper.getType() == ChatMessageType.SERVER)
```
Add line between variable and if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+			int index = i + INFO_START;
+
+			Widget textWidget = client.getWidget(WidgetID.MTA_ALCHEMY_GROUP_ID, index);
+			if (textWidget == null)
```
Add line between variable and if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+
+			Widget pointsWidget = client.getWidget(WidgetID.MTA_ALCHEMY_GROUP_ID, index + INFO_START);
+			int points = Integer.parseInt(pointsWidget.getText());
+			if (points == BEST_POINTS)
```
Add line between variable and if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+
+			WorldPoint mine = client.getLocalPlayer().getWorldLocation();
+			double objectDistance = object.getWorldLocation().distanceTo(mine);
+			if (nearest == null || distance > objectDistance)
```
Add line between variable and if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+	{
+		int start = AlchemyItem.indexOf(item);
+		int distance = 8 - index;
+		start = (start + distance) % 8;
```
Maybe declare distance first, and then int start = (AlchemyItem.indexOf(item)+ distance) % 8;

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+		int start = AlchemyItem.indexOf(item);
+		int distance = 8 - index;
+		start = (start + distance) % 8;
+		for (int i = 0; i < CAPACITY; i++)
```
Add line between variable and for loop

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+		for (int i = 0; i < cupboards.size(); i++)
+		{
+			GameObject cupboard = cupboards.get(i);
+			if (cupboard == null || i >= CAPACITY)
```
Add line between variable and if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+			}
+
+			AlchemyItem location = locations[i];
+			if (location == AlchemyItem.EMPTY)
```
Add line between variable and if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+				continue;
+
+			String text = location.toString();
+			if (text.equals(best))
```
Add line between variable and if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+			{
+				BufferedImage image = IMAGE_MAP.get(location);
+				Point canvasLoc = Perspective.getCanvasImageLocation(client, graphics, cupboard.getLocalLocation(), image, IMAGE_Z_OFFSET);
+				if (canvasLoc != null)
```
Add line between variable and if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+			else
+			{
+				Point canvasLoc = Perspective.getCanvasTextLocation(client, graphics, cupboard.getLocalLocation(), text, TEXT_Z_OFFSET);
+				if (canvasLoc != null)
```
Add line between variable and if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+			if (!suggest(suggest))
+			{
+				int index = (suggest + 3) % CAPACITY;
+				if (!suggest(index))
```
Add line between variable and if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+				{
+					ItemComposition composition = client.getItemDefinition(item.getId());
+					String temp = best;
+					if (temp.equals("Adamant Helm"))
```
Add line between variable and if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java:
```diff
@@ -0,0 +1,394 @@
+	private boolean suggest(int index)
+	{
+		AlchemyItem item = locations[index];
+		if (item == AlchemyItem.UNKNOWN && getConfig().mtaHintArrows())
```
Add line between variable and if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/EnchantmentRoom.java:
```diff
@@ -0,0 +1,177 @@
+				infoBoxManager.removeIf(e -> e.getPlugin() instanceof MTAPlugin);
+				this.counter = null;
+			}
+			return;
```
add line above

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/EnchantmentRoom.java:
```diff
@@ -0,0 +1,177 @@
+			infoBoxManager.addInfoBox(counter);
+		}
+
+
```
remove empty line

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/EnchantmentRoom.java:
```diff
@@ -0,0 +1,177 @@
+			setHiddenWidgets(true);
+		}
+
+
```
remove empty line

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/EnchantmentRoom.java:
```diff
@@ -0,0 +1,177 @@
+		Player local = client.getLocalPlayer();
+		WorldPoint nearest = null;
+		double min = Double.MAX_VALUE;
+		for (int x = 0; x < tiles[z].length; x++)
```
Add line between variable and for loop

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/EnchantmentRoom.java:
```diff
@@ -0,0 +1,177 @@
+			for (int y = 0; y < tiles[z][x].length; y++)
+			{
+				Tile tile = tiles[z][x][y];
+				if (tile == null || tile.getGroundItems() == null)
```
Add line between variable and if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/EnchantmentRoom.java:
```diff
@@ -0,0 +1,177 @@
+							min = dist;
+							nearest = tile.getWorldLocation();
+						}
+						count++;
```
add line above

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/EnchantmentRoom.java:
```diff
@@ -0,0 +1,177 @@
+				}
+			}
+		}
+		if (nearest != null && getConfig().mtaHintArrows())
```
add line above

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/EnchantmentRoom.java:
```diff
@@ -0,0 +1,177 @@
+	public boolean inside()
+	{
+		Widget inside = client.getWidget(WidgetID.MTA_ENCHANTMENT_GROUP_ID, 0);
+
```
remove empty line

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/EnchantmentRoom.java:
```diff
@@ -0,0 +1,177 @@
+	public void setHiddenWidgets(boolean hide)
+	{
+		Widget text = client.getWidget(WidgetInfo.MTA_ENCHANTMENT_BONUS_TEXT);
+		for (int i = IMAGE_START_ID; i < IMAGE_END_ID; i++)
```
Add line between variable and for loop

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/EnchantmentRoom.java:
```diff
@@ -0,0 +1,177 @@
+		for (int i = IMAGE_START_ID; i < IMAGE_END_ID; i++)
+		{
+			Widget image = client.getWidget(WidgetID.MTA_ENCHANTMENT_GROUP_ID, i);
+			if (image != null)
```
Add line between variable and if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/EnchantmentRoom.java:
```diff
@@ -0,0 +1,177 @@
+				image.setHidden(hide);
+			}
+		}
+		if (text != null)
```
add line above

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/GraveyardCounter.java:
```diff
@@ -0,0 +1,64 @@
+public class GraveyardCounter extends Counter
+{
+	private int count;
+
```
remove empty line

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/GraveyardRoom.java:
```diff
@@ -0,0 +1,198 @@
+		{
+			this.bonesImage = itemManager.getImage(ItemID.ANIMALS_BONES);
+		}
+		if (!inside() || !getConfig().graveyard())
```
add line above

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/GraveyardRoom.java:
```diff
@@ -0,0 +1,198 @@
+				infoBoxManager.removeIf(e -> e instanceof GraveyardCounter);
+				this.counter = null;
+			}
+			return;
```
add line above

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/GraveyardRoom.java:
```diff
@@ -0,0 +1,198 @@
+	public void itemContainerChanged(ItemContainerChanged event)
+	{
+		ItemContainer container = event.getItemContainer();
+		if (container == client.getItemContainer(InventoryID.INVENTORY))
```
Add line between variable and if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/GraveyardRoom.java:
```diff
@@ -0,0 +1,198 @@
+		{
+			this.items = container.getItems();
+			this.score = score();
+			if (this.counter != null)
```
Add line between variable and if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/GraveyardRoom.java:
```diff
@@ -0,0 +1,198 @@
+	private int score()
+	{
+		int score = 0;
+		for (Item item : items)
```
Add line between variable and for loop

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/GraveyardRoom.java:
```diff
@@ -0,0 +1,198 @@
+	private boolean isBone(int id)
+	{
+		int sub = id - BONES_START;
+
```
remove empty line

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/GraveyardRoom.java:
```diff
@@ -0,0 +1,198 @@
+		for (int i = INFO_WIDGET_START; i <= INFO_WIDGET_END; i++)
+		{
+			Widget image = client.getWidget(WidgetID.MTA_GRAVEYARD_GROUP_ID, i);
+			if (image != null)
```
Add line between variable and if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/GraveyardRoom.java:
```diff
@@ -0,0 +1,198 @@
+				image.setHidden(hide);
+			}
+		}
+
```
remove empty line

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAInventoryOverlay.java:
```diff
@@ -0,0 +1,60 @@
+	public Dimension render(Graphics2D graphics)
+	{
+		graphics.setFont(FontManager.getRunescapeBoldFont());
+		for (MTARoom room : plugin.getRooms())
```
Add line between variable and for loop

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAPlugin.java:
```diff
@@ -0,0 +1,130 @@
+@PluginDescriptor(name = "Mage Training Arena")
+public class MTAPlugin extends Plugin
+{
+
```
remove empty line

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAPlugin.java:
```diff
@@ -0,0 +1,130 @@
+	public void startUp()
+	{
+		this.rooms = new MTARoom[] {alchemyRoom, graveyardRoom, telekineticRoom, enchantmentRoom};
+		for (MTARoom room : rooms)
```
Add line between variable and if statement

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/MTARoom.java:
```diff
@@ -0,0 +1,54 @@
+
+	public void under(Graphics2D graphics2D)
+	{
+
```
remove empty line

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/MTARoom.java:
```diff
@@ -0,0 +1,54 @@
+
+	public void over(Graphics2D graphics2D)
+	{
+
```
remove empty line

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/MTASceneOverlay.java:
```diff
@@ -0,0 +1,61 @@
+	public Dimension render(Graphics2D graphics)
+	{
+		graphics.setFont(FontManager.getRunescapeFont());
+		for (MTARoom room : plugin.getRooms())
```
Add line between setfont and for loop

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/Maze.java:
```diff
@@ -0,0 +1,74 @@
+
+public enum Maze
+{
+
```
remove empty line

pbphl0w commented:
I have an idea for the Telekinetic room:

Sometimes the guardian only moves 2 tiles before you have to telegrab it again. For that reason, I would suggest that the tile that is highlighted is the tile that is closest to the corner of the next tile. That way there is not a lot of dead time (telegrab animation blocks movement).

I have created this paint image to demonstrate what I mean:
![image](https://user-images.githubusercontent.com/40008479/41068384-8154985e-69e9-11e8-93e6-5f8b3dd31e9c.png)

Toocanzs commented:
As @pbphl0w said, it's really important to show how far you can run while you are waiting for the guardian to stop moving as it saves you a bit of time each run. 

#3436 has this in it, and I don't particularly care if either this MTA plugin or yours is implemented, but it should certainty show this efficient path.

Jasper-ketelaar commented:
@Toocanzs @pbphl0w Right that makes sense didn't think of that, I'll go ahead and implement this; shouldn't really require any changes except for using the next tile in calculations too. Thanks for the suggestion.

Jasper-ketelaar commented:
## Update

Fixed @pbphl0w / @Toocanzs comment and cleared up the config a lot as suggested by @psikoi.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/mta/EnchantmentRoom.java:
```diff
@@ -0,0 +1,159 @@
+		WorldPoint nearest = null;
+		double min = Double.MAX_VALUE;
+
+		for (int x = 0; x < tiles[z].length; x++)
```
This should use the item layer changed event probably to find/store dragonstones

Abextm commented:
It seems I have broken it.
![2018-06-12_00-20-41](https://user-images.githubusercontent.com/12366911/41273398-8ba0cbd4-6dd6-11e8-9562-b513ad602d50.png)

Jasper-ketelaar commented:
Yeah I think I've done some stuff to the telekinetic room that makes it less robust. Working on rewriting the telekinetic structure in such a way that it's a lot more efficient and robust than it is in its current state.

Adam- commented:
I have cleaned up alch/enchant/graveyard. For some of them I did not like how you were hiding the widgets so I removed them for now. The WorldArea API class overlaps the telekinetic collision logic some, so I think it should be adjusted to use that.

LeviSchuck commented:
Well done on this Jasper

YepImRobbie commented:
Thank you very much for doing this!

YepImRobbie commented:
Could we have a number next to the banana one that states the qty of empty inventory slots I have? Would be much appreciated.

farcast3r commented:
Could we please have a 24 second timer for the enchantment room (when the bonus shapes change)?
